codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +520,49 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
                 try {
-                    connectionState().incrementStreamWindow(-writtenBytes);
-                    incrementStreamWindow(-writtenBytes);
+                    boolean writeError = true;
+                    try {
+                        writing = true;
```

Comment:
Ad 1) In what scenario would it make sense to do that?
Ad 2) no, the writing makes sure the queue can't be modified while a write is ongoing. cancel still errors the whole queue. It's in there to prevent a malicious implementation of FlowControlled.write to modify the queue.